feat: Support for Kubernetes (#1058)
GoogleCloudPlatform/professional-services-data-validator: README.md

}
```

### Running DVT with YAML Configuration Files

Running DVT with YAML configuration files is the recommended approach if:

> * you want to customize the configuration for any given validation by providing use case specific CLI arguments or editing the YAML configuration file OR

Reviewer: ```suggestion
* you want to customize the configuration for any given validation OR
```

Suggest making this simpler since it may be confusing to mention "CLI arguments" in this section

* you want to run DVT at scale (i.e. row validations across many partitions)

> Nearly every validation command can take the argument `-c <file-name>` (the `generate-table-partitions` command takes  `-cdir <directory-name>`) where one or more yaml files are produced. These yaml files can be modified (for more specific validations), see below

Reviewer: ```suggestion
Nearly every validation command can take the argument `-c <file-name>` (the `generate-table-partitions` command takes  `-cdir <directory-name>`) where one or more yaml files are produced. These yaml files can be modified for custom validations.
```

DVT validations can be run concurrently (horizontal scaling) using GKE (Kubernetes Jobs) or Cloud Run Jobs, GCP's managed Kubernetes offering. In order to run DVT in a container, you have to build a docker image, [see instructions](https://github.com/GoogleCloudPlatform/professional-services-data-validator/tree/develop/samples/docker#readme). Set the `PSO_DV_CONFIG` environment variable to point to a GCS prefix where the connection configuration files are stored. In Cloud Run you can use the option `--set-env-vars` or `--update-env-vars` to pass [the environment variable](https://cloud.google.com/run/docs/configuring/services/environment-variables#setting). We recommend that you use the `bq-result-handler` to save your validation results. In order to validate partitions concurrently, run DVT in Kubernetes or Cloud Run as shown below:
```
data-validation (--verbose or -v) (--log-level or -ll) configs run

>   [--kube-completions or -kc]

Reviewer: Let's add a description to the right of `-kc` flag similar to how we do in `-cdir` i.e "Indicate that validations are running in parallel Kubernetes jobs"

Also, let's clarify: should this flag should only be used for Kubernetes or for Cloud Run as well?

Author: Done

  --config-dir or -cdir GCS_DIRECTORY 
              where GCS_DIRECTORY = CONFIG_DIR/SOURCE_SCHEMA.SOURCE_TABLE, where CONFIG_DIR, SOURCE_SCHEMA and SOURCE_TABLE are
              the arguments provided to generate-table-partitions to generate the partition yamls. GCS_DIRECTORY is the directory
              where the partition files numbered 0000.yaml to <partno-1>.yaml are stored.
```

> The Cloud Run and Kubernetes pods must run in a network with access to the database servers. Every Cloud Run job is associated with a [service account](https://cloud.google.com/run/docs/securing/service-identity). You need to ensure that this service account has access to Google Cloud Storage (to read connection configuration and yaml files) and BigQuery (to publish results). If you are using Kubernetes, you will need to use a service account with the same privileges as mentioned for Cloud Run. In Kubernetes, you will need to set up [workload identity](https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity) so the DVT container can impersonate the service account.

Reviewer: Let's reference your docs/internal/kubernetes_jobs somewhere in this section.

Reviewer: We can mention in this section something like: "To read more, see docs/internal/kubernetes_jobs"

In Cloud Run, the [job](https://cloud.google.com/run/docs/create-jobs) must be run as multiple, independent tasks with the task count set to the number of partitions generated. In Kubernetes, set the number of completions to the number of partitions generated - see [Kubernetes Parallel Jobs](https://kubernetes.io/docs/concepts/workloads/controllers/job/#parallel-jobs). The option `--kube-completions or -kc` tells DVT that many DVT containers are running in a Kubernetes cluster. Each DVT container only validates the specific partition yaml (based on the index assigned by Kubernetes control plane). If the `-kc` option is used and you are not running in indexed mode, you will receive a warning and the container will process all the validations sequentially. If the `-kc` option is used and a config directory is not provided (a `--config-file` is provided instead), a warning is issued.

> By default, each partition validation is retried upto 3 times if there is any error. In Kubernetes and Cloud Run, you can set the parallelism to the number you want. If the parallelism is set to too high a number, then the databases being queried can become overloaded, resulting in timeouts and errors. If you are validating 1000's of partitions in parallel, you may find that setting the parallelism too high (say 100) may result in timeouts and slow down the validation.   

Reviewer: ```suggestion
By default, each partition validation is retried up to 3 times if there is an error. In Kubernetes and Cloud Run, you can set the parallelism to the number you want. Keep in mind that if you are validating 1000's of partitions in parallel, you may find that setting the parallelism too high (say 100) may result in timeouts and slow down the validation.   
```

> View the complete YAML file for a Grouped Column validation on the

Reviewer: Lines 469 - 472 should be moved up before the 'Scaling DVT to run 10's to 1000's of validations concurrently' section to line 449.

DVT validations can be run concurrently (horizontal scaling) using GKE (Kubernetes Jobs) or Cloud Run Jobs. In order to run DVT in a container, you have to build a docker image, [see instructions](https://github.com/GoogleCloudPlatform/professional-services-data-validator/tree/develop/samples/docker#readme). Set the `PSO_DV_CONFIG_HOME` environment variable to point to a GCS prefix where the connection configuration files are stored. In Cloud Run you can use the option `--set-env-vars` or `--update-env-vars` to pass [the environment variable](https://cloud.google.com/run/docs/configuring/services/environment-variables#setting). We recommend that you use the `bq-result-handler` to save your validation results. In order to validate partitions concurrently, run DVT in Kubernetes or Cloud Run as shown below:
```
data-validation (--verbose or -v) (--log-level or -ll) configs run

>   [--kube-completions or -kc]

Reviewer: ```suggestion
  [--kube-completions or -kc] Specifies that validation is being run in Kubernetes or Cloud Run in indexed completion mode.
```

When documenting the command itself, we should keep it short. The paragraphs below should link to the kubernetes docs so the user knows how to set the completions in GKE.
